diff --git a/cmds/cpud/serve.go b/cmds/cpud/serve.go
index f8ba5f7..aeb2e8f 100644
--- a/cmds/cpud/serve.go
+++ b/cmds/cpud/serve.go
@@ -9,10 +9,7 @@ import (
 	"log"
 	"math"
 	"net"
-	"os"
-	"os/exec"
 	"strconv"
-	"syscall"
 	"time"
 
 	// We use this ssh because it implements port redirection.
@@ -54,26 +51,7 @@ func initsetup() error {
 	if err := cpuSetup(); err != nil {
 		log.Printf("CPUD:CPU setup error with cpu running as init: %v", err)
 	}
-	cmds := [][]string{{"/bin/sh"}, {"/bbin/dhclient", "-v", "--retry", "1000"}}
-	verbose("Try to run %v", cmds)
-
-	for _, v := range cmds {
-		verbose("Let's try to run %v", v)
-		if _, err := os.Stat(v[0]); os.IsNotExist(err) {
-			verbose("it's not there")
-			continue
-		}
 
-		cmd := exec.Command(v[0], v[1:]...)
-		cmd.Stdin, cmd.Stdout, cmd.Stderr = os.Stdin, os.Stdout, os.Stderr
-		cmd.SysProcAttr = &syscall.SysProcAttr{Setctty: true, Setsid: true}
-		verbose("Run %v", cmd)
-		if err := cmd.Start(); err != nil {
-			verbose("Error starting %v: %v", v, err)
-			continue
-		}
-	}
-	verbose("Kicked off startup jobs, now serve cpu sessions")
 	return nil
 }
 
